src/Aspire.Hosting.Azure.Functions/AzureFunctionsProjectResourceExtensions.cs

```diff
@@ -0,0 +1,150 @@
+using Aspire.Hosting.ApplicationModel;
+using Aspire.Hosting;
+
+namespace Aspire.Hosting.Azure;
+
+/// <summary>
+/// Extension methods for <see cref="AzureFunctionsProjectResource"/>.
+/// </summary>
+public static class AzureFunctionsProjectResourceExtensions
+{
+    /// <summary>
+    /// Add an Azure Functions project to the distributed application.
+    /// </summary>
+    /// <typeparam name="TProject"></typeparam>
+    /// <param name="builder"></param>
+    /// <param name="name"></param>
+    /// <returns></returns>
+    public static IResourceBuilder<AzureFunctionsProjectResource> AddAzureFunctionsProject<TProject>(this IDistributedApplicationBuilder builder, string name) where TProject : IProjectMetadata, new()
+    {
+        var projectDirectory = Path.GetDirectoryName(new TProject().ProjectPath)!;
+        var resource = new AzureFunctionsProjectResource(name, "func", projectDirectory);
+
+        // Add the default storage resource if it doesn't already exist.
+        var storage = builder.Resources.OfType<AzureStorageResource>().FirstOrDefault(r => r.Name == "azure-functions-default-storage");
```

Comment:
Are we assuming that if a user wants to configure their own `AzureWebJobsStorage` they must configure it with this resource name for us to pick it up?

That seems like a sensible constraint to me but wanted to confirm that's what's goin on here.

Reply:
Nope, they just need to use `WithHostStorage` to do so. This default is only effective if they don't call that method (we keep this auto injected default)

Reply:
We can't avoid needing a name and a storage resource.

Reply:
I'm saying - instead of hard-coding a name `"azure-functions-default-storage"`. Would it be better if we created an internal Annotation (or similar) that only we could add to "mark" the AzureStorageResource as the default one? That way we don't rely on the Name, which is something that a user can manipulate.

Reply:
I don't understand how that would work. This resource needs to exist at some point and at that point it needs to have a name.

Reply:
> and at that point it needs to have a name.

That's fine, we can use some name like `"azure-functions-default-storage"` to give it a name.

I'm talking about how we recognize which AzureStorageResource is the default one. Basing that check on the name doesn't seem great. Imagine if a user wrote code like:

```C#
builder.AddAzureStorage("azure-functions-default-storage");

builder.AddAzureFunctionsProject<Projects.MyFuncs>("MyFuncs");
```

if they did that, this code would pick up the storage account they added on line 1. But instead we really want them to say:

```C#
var defaultStorage = builder.AddAzureStorage("azure-functions-default-storage");

builder.AddAzureFunctionsProject<Projects.MyFuncs>("MyFuncs")
    .WithHostStorage(defaultStorage);
```

Reply:
Ah, you want us to mark the default one. That's fine. If they did the above it would explode saying that they are multiple resources with the same name. Our best bet is to pick a name something that has a low likelyhood of collision

Reply:
I don't think the collisions are super likely to happen but I added this to https://github.com/dotnet/aspire/issues/5563 to track in a follow-up.